### ISSUE TITLE
Gravis Ultrasound (GUS) emulation

### ISSUE THREAD

**Issue by volkertb** [dosemu2/dosemu2#1370]
How feasible would it be to add support for emulating the Gravis Ultrasound (GUS) sound card as an (optional) alternative to SB16?

The distinct advantage of the GUS over Sound Blaster cards (and clones) was that it supported hardware mixing of multiple sound and music channels, which a number of DOS games actually took advantage of. Therefore, when running such games in dosemu2 with hypothetical GUS emulation, the mixing could be performed host-side in a separate thread, which could potentially yield better performance and audio fidelity.

Asking you to implement such emulation from scratch would be unreasonable. It would make more sense to "borrow" such emulation code from another project with a compatible license. The GUS emulation code in the dosbox-staging project (a Dosbox fork) might be particularly interesting, both quality- and feature-wise: https://github.com/dosbox-staging/dosbox-staging/wiki/Gravis-UltraSound-Enhancements

Thank you for considering this.

**Comment by volkertb** [comment]
Heh, that's funny. This issue is has number #1370, just like the Ensoniq AudioPCI ES1370 sound chip, which formed the heart of one of the first PCI sound cards back in the day. I know, I'm a nerd. :sweat_smile:

**Comment by stsp** [comment]
This is a low prio thing for us, as all the
games that support gus, also support sb16.
Of course there are gus-only demos.
Note that with sb16 you can mix pcm+fm+midi
at the same time.

**Comment by volkertb** [comment]
> Note that with sb16 you can mix pcm+fm+midi
at the same time.

Sorry for the confusion. That was not the kind of hardware mixing that I was referring to. The GUS also supports hardware mixing of multiple PCM channels (a.k.a. "multichannel" audio). Games that make use of that functionality (either for sound effects, tracker music or a combination of both) would potentially benefit from this, since a hypothetical GUS emulator in dosemu2 could perform such mixing in a separate thread. Games that supported this on the GUS would fall back to software mixing on SB16, but would often perform such software mixing at a lower frequency (and thus with a lower effective quality) than the GUS was capable of, so as not to tax the CPU too much.

**Comment by stsp** [comment]
Yes, was just pointing that even with sb16
you can already benefit from separate threads
and mixing, and in a much larger scale actually,
than that of GuS. Rendering midi is a very
CPU-expensive task, and doing it in another
thread is a huge win compared to just mixing
PCM samples.

So GuS is really a low prio thing. And its
quite frightening one, too, because it has
the SB16 emulation registers, so will likely
conflict with sb16 - they cannot be enabled
together I suppose. And its drivers need
a port hooking API that we do not have
(again, maybe only for sb16 emulation).
I think it will bring enormous amount of
problems for a very small benefit.

**Comment by jschwartzenberg** [comment]
I think it would also help to have a list of software that would work noticeably better with this.

**Comment by stsp** [comment]
I don't think such list will contain a single
entry. You can list a software that doesn't
produce any sound w/o GuS - that would
include a few demos.
But as for "noticeably better" - there is none,
because at best you can off-load the pcm
mixing to other threads (and you won't notice
that change)

**Comment by volkertb** [comment]
Typically, if a DOS game has Amiga-like tracker music as opposed of FM or MIDI music, and it prompts you to select a "music quality level" or the desired sampling frequency when you select something other than a GUS as your sound card in the game's setup program, it will almost certainly use the GUS hardware multichannel mixing capabilities.

I know of Jazz Jackrabbit, Epic Pinball and Pinball Fantasies as examples.

See also this thread on Vogons for a longer (probably still non-exhaustive) list: https://www.vogons.org/viewtopic.php?f=62&t=55274

**Comment by stsp** [comment]
It will, but this doesn't mean
"noticeably better" - you won't notice
anything.

**Comment by volkertb** [comment]
I'm not sure, but some games might only support software mixing at lower frequencies in order not to impact CPU utilization (and thus overall gameplay) too much. As a result, such games would actually sound better on a GUS (emulated or real).

This other Vogons thread has an even longer list of games that potentially benefit from GUS support: https://www.vogons.org/viewtopic.php?f=46&t=36661

**Comment by volkertb** [comment]
Don't worry, I'm not trying to convince you to drop everything to work on this right now. I agree that proper and full SB16 support would definitely suffice for the time being. I'm just debating the potential merits, even if it would apply to only a small number of games. 😉

At any rate, when it comes to the DOS gaming era, the only sound card that would potentially bring any added value when emulating, besides Sound Blaster and MPU-401, would be the GUS, both for those games that took advantage of it and for the GUS-only demos that you mentioned.

Although there are also some DOS games that supported stereo sound only on the Sound Blaster Pro, with the Sound Blaster 16 falling back to legacy Sound Blaster (mono) mode. But that is probably best discussed in a separate issue. 🙂

**Comment by stsp** [comment]
IIRC GuS was primarily beloved by its
midi synthesis. That's where it really made
the difference, but we already have GM over
MPU-401. :) For that to have on a real SB16,
one needed to buy WaveBlaster.

**Comment by volkertb** [comment]
It wasn't just the MIDI. The fact that you could upload any sample to its on-board RAM and use it as an instrument is what made it so popular with the demoscene in particular. Demo coders and also some game developers with roots in the  demoscene basically considered it a supercharged version of the Paula chip from the Amiga, and used it as such.

Although any wavetable sound card with RAM for uploadable patches or soundfonts could theoretically be used for mixing tracker music and multichannel digitized sound effects in hardware, I'm not aware of any non-GUS sound cards that were utilized in DOS games that way.

This is what made the GUS so unique IMO.

But of course, in games that supported either FM or General MIDI for music, the GUS indeed doesn't provide any added value in terms of emulation. I agree with that. We're explicitly talking about games with tracker music. (Although perhaps there were also some games with MIDI/FM music that still utilized the GUS for mixing sound effects in hardware.)

**Comment by jschwartzenberg** [comment]
https://invidious.kavin.rocks/watch?v=92olhbB3KKM&t=654&disable_polymer=true

Unfortunately he doesn't compare with an SB16.

**Comment by stsp** [comment]
This video says there was a GuS
version with just the midi synth:
https://en.wikipedia.org/wiki/Gravis_Ultrasound#UltraSound_ACE_(Audio_Card_Enhancer)
That speaks quite for itself.

**Comment by volkertb** [comment]
Yeah, the Ultrasound ACE by design lacked the on-board hardware necessary for Sound Blaster emulation, so it could coexist alongside an actual Sound Blaster or compatible card. It was released a few years later after the original GUS and marketed as a wavetable upgrade for people who already had a Sound Blaster. This was driven by market demand since at that point it had become clear that Sound Blaster and General MIDI had become the undisputed industry sound standards. But the aforementioned GUS-optimized games would still work with the ACE, since it was based on the same GF1 chip. The Sound Blaster emulation on the original GUS and even in the later GUS Max wasn't very good anyway. The FM emulation was terrible and I believe only original Sound Blaster DAC compatibility was supported, so no Sound Blaster Pro or 16 support, as far as I can remember.

**Comment by volkertb** [comment]
As for the the GUS Max, that was an improved version that had a dedicated chip (I believe the same Crystal or ESS chip as used in WSS compatible cards) for digital sound effects separately from the GF1 chip. I believe it could offer somewhat better Sound Blaster emulation that way, albeit still with crappy OPL2 FM emulation. But the original GUS used the GF1 chip for both music and digital sound effects.

**Comment by volkertb** [comment]
I deleted my last few comments, since they may have had some inaccuracies.

**Comment by volkertb** [comment]
But the ACE was released some years after the original GUS and marketed as a wavetable upgrade, since the market required Sound Blaster compatibility and the mediocre Sound Blaster emulation that the GUS provided didn't cut it. It was still based on the same GF1 chip as was used in the original GUS, and that single chip handled and mixed both music and digitized sound effects. It wasn't until the GUS Max that an addional chip was added that was dedicated to digital SFX playback.

As far as I know, the ACE was still compatible with the original GUS, at least in terms of playback (not recording).

**Comment by jschwartzenberg** [comment]
So it would make sense to just emulate the ACE?

**Comment by volkertb** [comment]
I believe so. Why bother with the recording part if the goal is to run games?

Also, I believe an emulated ACE would more easily coexist with an emulated Sound Blaster 16, since (if I'm not mistaken), the ACE lacked support for hardware-assisted Sound Blaster emulation, which was by design, since it was meant to coexist with actual Sound Blasters (or compatible cards). If I remember correctly, the hardware-assisted Sound Blaster emulation support (which would allow such emulation to work even in protected mode games) was what made the original GUS hog the typical Sound Blaster resources (I/O addresses, IRQ, DMA), and thus made it hard to coexist with other sound cards.

But frankly I would be surprised if the existing GUS emulation code in other open source projects (such as DOSBox Staging) included such things, because emulating the Sound Blaster emulation support of the GUS would be rather pointless, and games did not require recording support.

**Comment by stsp** [comment]
Even ACE is too much for us because
we already have gm (and can add mt32).
I suppose gus hase entirely different
midi interface, or is it still mpu-401?
So I wonder if we can only emulate the
gus dsp within our sb16 infrastructure,
mainly mpu-401 and sb16-mixer.
The thing I want to see least, is 2 mixers
running in parallel, with user changing volume
on sb16 mixer but nothing changes for
gus, and vice versa.

**Comment by jschwartzenberg** [comment]
> The thing I want to see least, is 2 mixers
> running in parallel, with user changing volume
> on sb16 mixer but nothing changes for
> gus, and vice versa.

But does this problem not exist with the GM support as well?

**Comment by stsp** [comment]
No such problem with gm.
Use MIDI scroller in sb16-mixer.

**Comment by jschwartzenberg** [comment]
But then you could hook up the line-in scroller in the sb16-mixer to the GUS ACE emulation similarly.

**Comment by stsp** [comment]
That's what I am talking about.
We could probably only emulate
gus dsp inside our sb16 infrastructure.
Maybe some mixer additions should
be added too.
But first, I wonder if ACE also has an
mpu-401 uart for midi, or something else.

**Comment by jschwartzenberg** [comment]
I suspect it is entirely based on sampling/mixing digital audio. I did not read anything about MIDI.

A possible check would be running some GUS games like Jazz Jackrabbit in DOSBox or PCem and compare how they sound compared to running them with SB16 instead.

**Comment by stsp** [comment]
> I suspect it is entirely based on sampling/mixing digital audio. I did not read anything about MIDI.

How about the wikipedia link I provided above?
```
Marketed as a competitor to Wave Blaster-compatible cards, it is supposed to be installed alongside a SoundBlaster Pro/16 card as a sample-based synthesis (marketed as 'wavetable synthesis') upgrade.
```

**Comment by jschwartzenberg** [comment]
There is nothing there about MIDI compatibility.

From what I am reading the only thing the card can do in hardware is digital sample mixing and anything MIDI that worked with the card was using the card's features to emulate MIDI in software. It seems also that is how ALSA provided MIDI support for it. Maybe it would be possible to look at those driver sources.
See also https://www.infania.net/misc/GRAVIS/

**Comment by stsp** [comment]
>  There is nothing there about MIDI compatibility.

What do you mean by that?

> card was using the card's features to emulate MIDI in software.

What does it mean, use (hardware) card's features
to emulate MIDI in software?
Could you be more specific please?

**Comment by jschwartzenberg** [comment]
I mean there is no MIDI support whatsoever from a hardware point of view. I think this article explains it: https://nerdlypleasures.blogspot.com/2012/07/gravis-ultrasound-and-dos-games.html
> The card I have is the Gravis Ultrasound Audio Card Enhancer (ACE), v1.0.  This is a stripped down card designed to work with a Sound Blaster.  It has the no midi/gameport, no recording capabilities, and no CD-ROM/IDE interfaces.

**Comment by stsp** [comment]
Game port doesn't matter.
Even though sb16 IIRC had one,
the WaveBlaster daughterboard didn't
use it. It used an entirely different connector,
which allowed to get rendered PCM stream
back to SB16, so you could control it with
sb16-mixer. Obviously you couldn't control
the volume of a midi through the game port.

So... the lack of the game port means purely
nothing about a MIDI support.
Also please note that the software midi
rendering was absolutely impossible on
CPUs of that era.

**Comment by jschwartzenberg** [comment]
> Also please note that the software midi
> rendering was absolutely impossible on
> CPUs of that era.

I think this is where the GUS hardware features step in. This `ultramid` or `Mega-Em` could use them to handle specifically midi, but the hardware was abstracted to the point of mixing many digital samples. Maybe @volkertb knows more.

**Comment by stsp** [comment]
Well, wikipedia says rather clearly that
`Marketed as a competitor to Wave Blaster-compatible cards`.
And its quite clear what WaveBlaster is
and how it works.
The only question that I would raise,
is whether ACE had a DSP at all, as
WaveBlaster obviously did not.
The fact that it had midi synth, looks
indisputable for me.

**Comment by volkertb** [comment]
Okay, let me take a step back in describing this.

The GF1 chip (used in earlier Ultrasound models) and the AMD Interwave chip (used in later Ultrasound models) allowed for the mixing of multiple PCM channels, up to 32 channels in the case of the GF1. Each PCM channel could be played back at different frequencies, which could be used to control the pitch of the sample. Again, this is inspired by the Paula chip in the Amiga computers, but way more advanced, since the Paula chip could only mix 4 channels (2 left, 2 right).

Also, I believe each channel could individually be panned left, right or center.

This mixing of multiple PCM channels with variable pitch and pan per channel allowed for a number of things:

* Mixing tracker music (MOD files and such) in hardware
* Mixing digital sound effects as well as music in games in hardware (someone talking, with an explosion in the background, and the engine of a running car engine at the same time, and background music combined with that, etc)
* Allowing for General MIDI playback

The way the GUS supported General MIDI was by loading the instruments ("patches") into the card's RAM, just like the samples of a MOD file would be, and then playing them back. Basically, it would play back MIDI files just as it would play back tracker files (MOD, S3M, etc). The only difference would be that the instruments would be loaded from a standard General MIDI compatible patch set, as opposed to being bundled in the music file.

Now the GUS did not offer any hardware-level MPU-401 compatibility, neither UART mode nor intelligent mode. Of course, Gravis was initially hoping that the card would become popular enough that it would be supported by games natively. However, as a fallback, for games that didn't support the GUS, software was provided for emulating Sound Blaster with terrible FM (with a TSR called SBOS), as well as software that would emulate a Sound Blaster DSP with an MPU-401 with either a General MIDI or a Roland MT-32 synthesizer connected to it (switchable at the command-line). The difference between GM mode and MT-32 mode being merely a different mapping of instruments. Needless to say, the MT-32 mode sounded bad in a lot of games, since many games programmed the MT-32 to use custom instruments.

Alternatively to MPU-401 emulation, many games that "natively" supported the GUS, did so using DOS sound drivers, such as the Miles Audio Interface Library (AIL), MIDPAK/DIGPAK, HMI, etc. So Gravis would develop such drivers along with patching utilities to add native support to games that used such drivers, but didn't initially ship with an Ultrasound driver. Such games wouldn't program an MPU-401 interface directly, but they would often still standardize on General MIDI for how their music was composed. So Gravis had another tool called Ultramid, which basically loaded the General MIDI instruments ("patches") with the proper mappings into the RAM of the card. You'd then have to run Ultramid before running such a game. The GUS driver would then access the GUS directly, assuming the proper instruments were loaded with the right mappings, and play them (again, like playing MOD files, but with the instruments loaded separately in advance).

Anyway, a few years after the initial release of the GUS, Gravis saw how they had failed to make a dent in the dominance of Sound Blaster as a sound card standard. They also saw that a market was forming for Waveblaster upgrades for people with a Sound Blaster 16 (or clone). They already had their existing GUS design, perhaps they had a warehouse full of surplus GF1 chips at that point (this is speculation), and at that point more games already supported the GUS, also for MIDI music, but through Ultramid, not any MPU-401 emulation. So they developed a card without any stuff on it that would interfere with an actual Sound Blaster card (so no game port, and I assume no hogging of Sound Blaster resources for hardware-assisted emulation, etc), and marketed it as an alternative to Waveblaster upgrade.

With the emphasis on *alternative*. It wouldn't plug into a Waveblaster port on a Sound Blaster or anything, it would still be a card that would go into a separate ISA slot. With the included drivers and emulators (Ultramid and I assume a version of MegaEM without the Sound Blaster DSP emulation part), it would indeed make a decent alternative.

And indeed, when used only with games that used it for playing back General MIDI music, the GUS (whether the original version, the ACE or any later version or edition) would indeed not have any added value emulation-wise.

But like I said before, the GF1 and Interwave chips were multichannel PCM mixers that could handle music as wel as digital effects, or a combination thereof.

The ACE did not have a separate DSP, and neither did the original GUS. The GF1 chip handled and mixed both music and digital sound effects, and actually didn't make any distinction between them. It wasn't until the GUS Max that a separate chip for digital sound effects was added.

The original GUS (wich just the GF1 chip) had a 16-bit DMA channel as well as (expandable) on-board RAM. My guess is that it could mix sounds from the main RAM and from the card's RAM interchangably.

TL;DR: the GF1 chip in the GUS and the GUS ACE was a versatile beast that handled both music and digital sound effects, and could also be used for General MIDI playback because of its flexibility, combined with the right drivers and/or emulation TSRs.

**Comment by jschwartzenberg** [comment]
I don't believe that "marketed as" bears any relationship to how the hardware truly worked :)
The idea seems to have been that it was similarly sold as an add-on board.

Driver sources are here:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/sound/isa/gus

**Comment by volkertb** [comment]
Indeed. It was an ISA card of its own, albeit with fewer I/O connectors. Those of you who don't believe this, just do a search on Google Images for "Gravis UItrasound ACE".

**Comment by stsp** [comment]
Well, there is only one "unbeliever" in this thread. :)

So could you please clarify:
- GUS didn't have native mpu-401, but it could be
emulated together with sb16. And ACE didn't have
sb16 emulation. Is it logical to assume it didn't have
any mpu-401 support at all then? That would be
quite a disadvantage compared to WaveBlaster.
- `The ACE did not have a separate DSP, and neither did the original GUS.`
could you please clarify the consequences of that
for the digital sound output? Did ACE have PCM support,
or just midi?

**Comment by jschwartzenberg** [comment]
>     * GUS didn't have native mpu-401, but it could be
>       emulated together with sb16. And ACE didn't have
>       sb16 emulation. Is it logical to assume it didn't have
>       any mpu-401 support at all then? That would be
>       quite a disadvantage compared to WaveBlaster.

What I understand: not in hardware, only in software.


>     * `The ACE did not have a separate DSP, and neither did the original GUS.`
>       could you please clarify the consequences of that
>       for the digital sound output? Did ACE have PCM support,
>       or just midi?

From this "The GF1 chip handled and mixed both music and digital sound effects, and actually didn't make any distinction between them. It wasn't until the GUS Max that a separate chip for digital sound effects was added." I understand one chip handled both PCM & MIDI. So emulating that GF1 chip would provide ACE emulation. It would enable full GUS support as the non-ACE features are already covered by the SB emulation.

**Comment by stsp** [comment]
> What I understand: not in hardware, only in software.

You can't easily emulate this in software,
not with IO port trapping features of emm/qemm,
so no.

**Comment by volkertb** [comment]
> > What I understand: not in hardware, only in software.
> 
> You can't easily emulate this in software,
> not with IO port trapping features of emm/qemm,
> so no.

The Gravis Ultrasound supported some form of hardware-assisted emulation of Sound Blaster (and in later PCB revisions also MPU-401) cards. I believe it accomplished this by actually wiring itself up to Sound Blaster and MPU-401 resources (I/O addresses, IRQ, DMA), leaving the rest up to software emulation. This is how SBOS (the Sound Blaster emulation TSR that came with the GUS) could also work with protected mode games.

But of course this particular GUS feature doesn't have to be emulated, since dosemu2 is already emulating a SB16, and with much higher quality OPL3/FM emulation than the crappy FM emulation that SBOS provided.

Again, the only interesting use case for GUS emulation is for games and demos that...

1. ... used the GF1 or Interwave chip on GUS cards to mix multi-channel sound effects in hardware (allowing dosemu2 to alleviate the emulation thread by handling GUS/GF1 emulation in a separate thread)
2. ... used Amiga-like tracker music (as opposed to FM/MT-32/GM music) and utilized the GF1 or Interwave chip on GUS cards to mix the music channels of said tracker music in hardware.
3. Games that did both 1 and 2

Note: not all games that used tracker music made use of the GUS this way. Some such games would still do the mixing in software, in which case GUS emulation wouldn't add any performance benefits.

But for the games that did make use of the GUS this way, implementing GUS emulation would add the following benefits:

1. Better use of multi-core CPUs on host systems, since the software mixing (whether tracker music, digital sound effects, or both) could be taken care of in a separate thread
2. Potentially higher sound quality in at least some of such games, since the fallback to software mixing on Sound Blaster cards would often be done at a lower sample rate than when the hardware-mixing capability of GUS were utilized (for performance reasons)

**Comment by stsp** [comment]
There is no any multi-core benefit here.
I moved midi and fm renderings to the
separate threads because _they_ are
resource-intensive.
But mixing 2 (or 32) PCM stream is not
resource-intensive: its just a mere addition
of all values. Moreover, you can't separate
them, because you need all of them at
once to get the final value. There is just
nothing to parallelize.
Well if they have different sample rates,
you can resample them in a separate
threads, and then mix in a single thread.
But even this will probably be ineffective
because you will have to do a lot of buffering,
which adds latencies. More efficient approach
is to resample in the same thread where
mixing is done - which is what we do now.

**Comment by volkertb** [comment]
[This Vogons thread](https://www.vogons.org/viewtopic.php?f=46&t=36661) provides some interesting revelations on the topic of actual GUS hardware mixing support in games.

This quoted email from a sound developer who removed GUS hardware mixing support in version 1.666 of Doom was particularly sobering:

> All of the sound effects are now mixed in software, rather than on the GUS hardware. Why, you ask? Because of several reasons. First, is that the GF1 chip has a minimal ramp time that is much to long for very sharp effects. Second, because loading of the MUSIC patches uses all of the GUS memory, I had to DMA all eight sound effects to the card when played. This intern exposed a bug in the GF1 chip that Gravis did not find until my code started to beat on it. The bug would cause the bus to freeze and any program with it. The workaround is to keep DMA activities to a minimum by mixing in software and transfering only 1 channel to the GUS. But since the GF1 can't support auto-initialize DMA, and because the only way to play interleaved data on the card is to set two voices pointing into a single patch and setting the frequency so the every other sample is skipped, you don't get the benifit of sample smoothing from the GF1.
> 
> Sorry, but that's the way it has to be 🙁
> 
> Paul Radek
> Digital Expressions, Inc.

Yeah, as disappointing as that is, I guess that does confirm that there is little added value in adding hardware-level GUS emulation to dosemu2. It ultimately only benefits demos and a handful of games.

At the end of the day, the vast majority of DOS games that supported the GUS natively basically treated it as a General MIDI synthesizer and a single-channel DMA-driven stereo DAC, completely forgoing its hardware mixing abilities for sound effects.

I'll just close this issue, then. :disappointed:

**Comment by stsp** [comment]
In addition, I think GUS has a very
fragile and undocumented features
of generating NMIs to do a HW-assisted
emulation of SB16.
And also, now, unlike in the beginning
of this thread, we DO have an mt32
emulation, so GUS really can't add
anything new for us any longer.
Its really for dosbox-X to implement
such things. And the last thing we need
is to compete with dosbox-X. :)
